src/app/clusters/door-lock-server/door-lock-server.cpp

```diff
@@ -3966,3 +4114,120 @@ void DoorLockServer::DoorLockOnAutoRelockCallback(System::Layer *, void * callba
+{
+    VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INCORRECT_STATE, ChipLogError(Zcl, "Delegate is nullptr"));
+
+    return aEncoder.EncodeList([delegate](const auto & encoder) -> CHIP_ERROR {
+        for (uint8_t i = 0; true; i++)
+        {
+            uint8_t buffer[kProtocolVersionMaxSize];
+            MutableByteSpan protocolVersion(buffer);
+            auto err = delegate->GetAliroExpeditedTransactionSupportedProtocolVersionAtIndex(i, protocolVersion);
+            if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+            {
+                return CHIP_NO_ERROR;
+            }
+            ReturnErrorOnFailure(err);
+            ReturnErrorOnFailure(encoder.Encode(protocolVersion));
+        }
+    });
+}
+
+CHIP_ERROR DoorLockServer::ReadAliroSupportedBLEUWBProtocolVersions(const ConcreteReadAttributePath & aPath,
+                                                                    AttributeValueEncoder & aEncoder, Delegate * delegate)
+{
+    VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INCORRECT_STATE, ChipLogError(Zcl, "Delegate is nullptr"));
+
+    return aEncoder.EncodeList([delegate](const auto & encoder) -> CHIP_ERROR {
+        for (uint8_t i = 0; true; i++)
+        {
+            uint8_t buffer[kProtocolVersionMaxSize];
+            MutableByteSpan protocolVersion(buffer);
+            auto err = delegate->GetAliroSupportedBLEUWBProtocolVersionAtIndex(i, protocolVersion);
+            if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+            {
+                return CHIP_NO_ERROR;
+            }
+            ReturnErrorOnFailure(err);
+            ReturnErrorOnFailure(encoder.Encode(protocolVersion));
+        }
+    });
+}
+
+CHIP_ERROR DoorLockServer::ReadAliroByteSpanAttribute(CHIP_ERROR (Delegate::*func)(MutableByteSpan &), MutableByteSpan & data,
+                                                      Delegate * delegate, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INCORRECT_STATE, ChipLogError(Zcl, "Delegate is nullptr"));
+
+    ReturnErrorOnFailure((delegate->*func)(data));
+    ReturnErrorOnFailure(aEncoder.Encode(data));
+    return CHIP_NO_ERROR;
+}
+
+// Implements the read functionality for the AttributeAccessInterface.
+CHIP_ERROR DoorLockServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    if (aPath.mClusterId != Clusters::DoorLock::Id)
+    {
+        // We shouldn't have been called at all.
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    Delegate * delegate = GetDelegate(aPath.mEndpointId);
+
+    switch (aPath.mAttributeId)
+    {
+    case AliroReaderVerificationKey::Id: {
+        uint8_t buffer[kAliroReaderVerificationKeyMaxSize];
+        MutableByteSpan aliroReaderVerificationKey(buffer);
+        return ReadAliroByteSpanAttribute(&Delegate::GetAliroReaderVerificationKey, aliroReaderVerificationKey, delegate, aEncoder);
```

Comment:
@bzbarsky-apple Ideally would have liked to pass in the size of the MutableByteSpan and created the MutableByteSpan inside ReadAliroByteSpanAttribute but i can't define dynamic array size for the uint8_t buffer (getting compilation error supported in c99). Any suggestions?

Reply:
You could template ReadAliroByteSpanAttribute on the size.   Whether that's more code or less code than what you have here, I don't know.  I also don't know whether the complexity of the pointer-to-member and separate function thing is worth it.  Was it a significant codesize savings?

Alternately, you could pass in the size, have the _buffer_ sized to the max of the possible sized in the callee, but have the _ByteSpan_ use the provided size.

Reply:
i did think about the latter - " have the buffer sized to the max of the possible sized in the callee, but have the ByteSpan use the provided size.". However this is not saving much with respect to code size. so i will leave it for now.